### ISSUE TITLE
About motion data

### ISSUE THREAD

**Issue by lucky9-cyou** [hanyangclarence/UniMuMo#3]
Thanks for your awesome work.

I want to convert your output to bvh, but I found that your joints are not in the smpl format. Do you have any suggestions for converting your joints or motion vectors to the bvh format?

**Comment by hanyangclarence** [comment]
Hi,

The visualization part was not mainly done by me so I'm not that familiar with the details of these formats and conversions. We convert our joint format to smpl by adapting the code in https://github.com/blackgold3/SemanticBoost `SMPLX/visualize_joint2smpl/simplify_loc2rot.py` . You may refer to this.

Or also, here is our modified code for your reference. The code is ugly, not written by me, but should work. 
```python
import numpy as np
import os
import torch
from SMPLX import smplx
import h5py
from SMPLX.visualize_joint2smpl.joints2smpl.src.smplify import SMPLify3D
from tqdm import tqdm
import argparse
import pickle
import pdb

class joints2smpl:

    def __init__(self, num_frames, device, model_path=None, json_dict=None):
        self.smpl_dir = model_path
        self.device = device
        # self.device = torch.device("cpu")
        self.batch_size = num_frames
        self.num_joints = 22  # for HumanML3D
        self.joint_category = "AMASS"
        self.num_smplify_iters = 30
        self.fix_foot = False
        
        smplmodel = smplx.create(self.smpl_dir, model_type="smpl", gender="neutral", ext="pkl",
                                 batch_size=self.batch_size).to(self.device)

        # ## --- load the mean pose as original ----
        # smpl_mean_file = os.path.join(json_dict["joints2smpl"], "neutral_smpl_mean_params.h5")
        smpl_mean_file = '/mnt/d/local_documentation/Render/SemanticBoost/SMPLX/visualize_joint2smpl/joints2smpl/smpl_models/neutral_smpl_mean_params.h5'
        file = h5py.File(smpl_mean_file, 'r')
        self.init_mean_pose = torch.from_numpy(file['pose'][:]).unsqueeze(0).repeat(self.batch_size, 1).float().to(self.device)
        self.init_mean_shape = torch.from_numpy(file['shape'][:]).unsqueeze(0).repeat(self.batch_size, 1).float().to(self.device)
        self.cam_trans_zero = torch.Tensor([0.0, 0.0, 0.0]).unsqueeze(0).to(self.device)
        #

        # # #-------------initialize SMPLify
        self.smplify = SMPLify3D(smplxmodel=smplmodel,
                            joints_category=self.joint_category,
                            num_iters=self.num_smplify_iters,
                            device=self.device)


    def npy2smpl(self, npy_path):
        out_path = npy_path.replace('.npy', '_rot.pkl')
        path_help = '/mnt/d/local_documentation/Render/SemanticBoost/SMPLX/visualize_joint2smpl'
        # joint path_help and out_path
        out_path = os.path.join(path_help, out_path)
        motions = np.load(npy_path, allow_pickle=True)[None][0]
        # print_batch('', motions)
        # n_samples = motions['motion'].shape[0]
        # n_samples = motions.shape[0]
        n_samples = 1
        all_thetas = []
        # for sample_i in tqdm(range(n_samples)):
            # thetas, _ = self.joint2smpl(motions['motion'][sample_i].transpose(2, 0, 1))  # [nframes, njoints, 3]
        thetas, smpl_trans = self.joint2smpl(motions)  # [nframes, njoints, 3]
        all_thetas.append(thetas)
        # motions['motion'] = np.concatenate(all_thetas, axis=0)
        motions = np.concatenate(all_thetas, axis=0)
        # print('motions', motions['motion'].shape)
        smpl_scaling = np.array([1])
        # switch x and z axis
        smpl_trans = smpl_trans[:, [1, 2 ,0 ]]
        # let x axis multiply -1
        # smpl_trans = smpl_trans * np.array([-1, 1, 1])
        # smpl_trans = np.zeros((200, 3))

        # store the data
        data_to_save = {
            "smpl_poses": motions,
            "smpl_scaling": smpl_scaling,
            "smpl_trans": smpl_trans
        }

        with open(out_path, 'wb') as f:
            pickle.dump(data_to_save, f)



    def joint2smpl(self, input_joints, init_params=None):
        if len(input_joints.shape) == 2:
            input_joints = input_joints.reshape(input_joints.shape[0], -1, 3)

        pred_pose = torch.zeros(self.batch_size, 72).to(self.device)
        pred_betas = torch.zeros(self.batch_size, 10).to(self.device)
        pred_cam_t = torch.zeros(self.batch_size, 3).to(self.device)
        keypoints_3d = torch.zeros(self.batch_size, self.num_joints, 3).to(self.device)

        # joints3d = input_joints[idx]  # *1.2 #scale problem [check first]
        keypoints_3d = torch.Tensor(input_joints).to(self.device).float()
        
        root_loc = torch.tensor(keypoints_3d[:, 0:1])   #### N * 1 * 3
        root_loc = root_loc - root_loc[[0], :, :]   ### N * 1 * 3
        root_loc = root_loc.squeeze(1).detach().cpu().numpy()
        # if idx == 0:
        if init_params is None:
            pred_betas = self.init_mean_shape
            pred_pose = self.init_mean_pose
            pred_cam_t = self.cam_trans_zero
        else:
            pred_betas = init_params['betas']
            pred_pose = init_params['pose']
            pred_cam_t = init_params['cam']

        if self.joint_category == "AMASS":
            confidence_input = torch.ones(self.num_joints)
            # make sure the foot and ankle
            if self.fix_foot == True:
                confidence_input[7] = 1.5
                confidence_input[8] = 1.5
                confidence_input[10] = 1.5
                confidence_input[11] = 1.5
        else:
            print("Such category not settle down!")

        new_opt_pose = self.smplify(
            pred_pose.detach(),
            pred_betas.detach(),
            pred_cam_t.detach(),
            keypoints_3d,
            conf_3d=confidence_input.to(self.device),
        )

        thetas = new_opt_pose.reshape(self.batch_size, 24 * 3)
        vecs = thetas.detach().cpu().numpy()

        return vecs, root_loc


if __name__ == '__main__':
    folder_list = ["/mnt/d/local_documentation/demo/motion2text/"]
    parser = argparse.ArgumentParser()
    parser.add_argument("--cuda", type=bool, default=True, help='')
    parser.add_argument("--device", type=int, default=0, help='')
    params = parser.parse_args()
    model_path = '/mnt/d/local_documentation/Render/SemanticBoost/SMPLX/visualize_joint2smpl/joints2smpl/smpl_models/SMPL_MALE.pkl' # TODO
    # simplify = joints2smpl(device_id=params.device, cuda=params.cuda)
    simplify = joints2smpl(600, params.device,model_path)

    for folder in folder_list:
        assert os.path.isdir(folder), f"Folder {folder} does not exist"
    file_list = []
    for folder in folder_list:
        file_list.extend([os.path.join(folder, f) for f in os.listdir(folder) if f.endswith('.npy')])

    for i, f in enumerate(file_list):
        data = np.load(f)
        data_length = data.shape[0]
        print(f'Processing {i+1}/{len(file_list)}: {f} with data length {data_length}')

        if data_length == 600:
            simplify.npy2smpl(f)
        else:
            print(f'Creating model for data length {data_length}')
            simplify_tempt = joints2smpl(data_length, params.device,model_path)
            simplify_tempt.npy2smpl(f)

```

**Comment by lucky9-cyou** [comment]
Thank you very much for your quick response, but I have encountered some problems. I wonder if you have encountered them as well:
![CleanShot 2024-11-23 at 03 19 40@2x](https://github.com/user-attachments/assets/5af096cf-b7c8-40af-8ebd-8239d28d7368)
I just load a numpy with shape [600, 22, 3]
And the loss of lbfgs is always `nan`:
![CleanShot 2024-11-23 at 03 25 14@2x](https://github.com/user-attachments/assets/8a3f36b3-666f-4b44-9b83-8aa46123178d)

**Comment by lucky9-cyou** [comment]
I just save the np file using following code:
![CleanShot 2024-11-23 at 03 27 34@2x](https://github.com/user-attachments/assets/aa30686e-375d-479a-a8f9-ef33e3b98638)

**Comment by lucky9-cyou** [comment]
I just use following command to generate joints:
```bash
python generate.py --ckpt ckpts/full.ckpt -t mo -mu_p ./data/music/audios/Counting_Stars.wav
```
The input and output files can be found at the following link:
https://drive.google.com/file/d/1C9JmFP8_ekEbiPd30vyhMX8WtO2F66jQ/view?usp=sharing

Thank you once again for your excellent work and help.

**Comment by hanyangclarence** [comment]
Oh I remembered that the person in charge of this part told me that, this conversion needs another model that requires extra training. So I guess a checkpoint might be needed. In my provided `SMPLX/visualize_joint2smpl/simplify_loc2rot.py` line 138, that path is the ckeckpoint. 

Here you can download the raw code. 
https://drive.google.com/file/d/1AxzrLv0bPAUIAEji57EMk2iRrnaIf2Mr/view?usp=drive_link
It's a buggy version, but you can use the model checkpoint in SMPLX\visualize_joint2smpl\joints2smpl\smpl_models\SMPL_MALE.pkl.

According to my github modification log, in addition to the modified `simplify_loc2rot.py` and the new `SMPL_MALE.pkl` in `SMPLX\visualize_joint2smpl\joints2smpl\smpl_models\`, the only modification is `SMPLX\visualize_joint2smpl\joints2smpl\src\prior.py` line 120, which I change to `full_gmm_fn = '/mnt/d/local_documentation/Render/SemanticBoost/SMPLX/visualize_joint2smpl/joints2smpl/smpl_models/gmm_08.pkl'`

Can you try these and see whether the problem solved?

**Comment by lucky9-cyou** [comment]
Thanks for your help very much. It is caused by pytorch version:
https://github.com/pytorch/pytorch/issues/131398
If I use pytorch 2.5.1 or 2.4.0, it will return nan. But if i use pytorch 2.0.1, it will be fine.

**Comment by hanyangclarence** [comment]
Okay😂that would be great!
